statsmodels/multivariate/manova.py

```diff
@@ -0,0 +1,346 @@
+
+    For y = x * params, where y is dependent variables and x is independent
+    variables, testing L * params * M = 0 where L is the contast matrix for
+    hypothesis testing and M is the transformation matrix for transforming the
+    dependent variables in y.
+
+    Algorithm:
+        H = (L * u * M)' * (L * u * M)
+        E = M' * sscpr * M
+        solve (H + E) * T = H
+    And then solving the eigenvalues of T * H
+
+    .. [1] ftp://public.dhe.ibm.com/software/analytics/spss/documentation/statistics/20.0/en/client/Manuals/IBM_SPSS_Statistics_Algorithms.pdf
+
+    Parameters
+    ----------
+    results : tuple
+        Output of ``fit_manova``
+    contrast_L : array-like
+        Contrast matrix for hypothesis testing. Each row is an hypothesis and
+        each column is an independent variable.
+        At least 1 row (1 by k_exog, the number of independent variables)
+    transform_M : array-like
+        Transform matrix. Default to be k_endog by k_endog identity
+        matrix (i.e. do not transform y matrix).
+
+    Returns
+    -------
+    results : MANOVAResults
+
+    """
+    df_resid, u, sscpr = results
+    M = transform_M
+    if M is None:
+        M = np.eye(u.shape[1])
+    L = contrast_L
+    t1 = L .dot(u)
+    q = matrix_rank(t1)
+    t1 = t1.dot(M)
+    H = t1.T.dot(t1)
+
+    # E = M'(Y'Y - B'(X'X)B)M
+    E = M.T.dot(sscpr).dot(M)
+
+    EH = np.add(E, H)
+    p = matrix_rank(EH)
+
+    # eigenvalues of (E + H)^H
+    eigv2 = np.sort(eigvals(solve(EH, H)))
+    return multivariate_stats(eigv2, p, q, df_resid)
+
+
+class MANOVA(Model):
+    """
+    Multivariate analysis of variance
+
+
+    Parameters
+    ----------
+    endog : array-like
+        Dependent variables (DV). A n_sample x n_y_var array where n_sample is
+        the number of observations and n_y_var is the number of DV.
+
+    exog : array-like
+        Independent variables (IV). A n_sample x n_x_var array where n is the
+        number of observations and n_x_var is the number of IV. An intercept is
+        not included by default and should be added by the user (models
+        specified using a formula include an intercept by default)
+
+    .. [1] ftp://public.dhe.ibm.com/software/analytics/spss/documentation/statistics/20.0/en/client/Manuals/IBM_SPSS_Statistics_Algorithms.pdf
+
+    Attributes
+    -----------
+    df_resid : float
+        The number of observation `n` minus the number of IV `q`.
+    sscpr : array
+        Sums of squares and cross-products of residuals
+    endog : array
+        See Parameters.
+    exog : array
+        See Parameters.
+    design_info : patsy.DesignInfo
+        Contain design info for the independent variables if model is
+        constructed using `from_formula`
+
+    """
+    def __init__(self, endog, exog, design_info=None, **kwargs):
+        self.design_info = design_info
+        self.fittedmod = fit_manova(exog, endog)
+        super(MANOVA, self).__init__(endog, exog)
```

Comment:
I was confused here. The issue for missing handling is if the user directly provides the design matrices for endog and exog, e.g. as DataFrames.
I had use the from_formula example from the unit tests, and there by default nan rows are removed. In this case we force it to a balanced multivariate dataset.

Initially I realized that fit_manova doesn't go through the data handling because the last return item of fit_manova is a DataFrame which shouldn't happen internally because internally everything is converted to numpy arrays (i.e. self.endog and self.exog are numpy arrays, and the pandas index are only available in the mod.data.orig_endog and mod.data.orig_exog.